### DISCUSSION TITLE
Wondering how to overclock rp2040 to 240mhz on platformio, and looking for advice on speeding up file reading.

### DISCUSSION THREAD

**Discussion by ws2bhhh** [earlephilhower/arduino-pico#1618] [Q&A] [answered]
Firstly, I write this at the top of main.cpp:
```
#undefine PICO_FLASH_SPI_CLKDIV
#define PICO_FLASH_SPI_CLKDIV 4
```
Then modify platfomio.ini:
```
board_build.f_cpu = 240000000L
board_build.flash_mode = qio
board_build.f_flash = 80000000L
```
But it doesn't work. No serial port appears at all on my computer. Same the Blink test. I'm sure my rp2040 can be overclocked to 240mhz cause I overclocked it successfully using microropython's machine.freq(240000000).

In terms of file reading, littlefs is too slow for me. It takes about 20ms to read a 1kb file, while it takes me only 7ms to read the same file using fatfs on esp32c3@160mhz. My programme is quite time sensitive, and reading a file should take at least less than 7ms.
```
file.read(buffer, file.size());
```
Is it possible to use SDFS on the on-board flash?

**Comment by ws2bhhh** [comment]
First problem has been solved. I think I may have bought a one-to-one copy of the YD RP2040. Also the configuration of the YD RP2040 seems to be wrong, the YD RP2040 doesn't come in a version with 2MB of flash, and the interface isn't SPI, but QSPI. (Maybe this configuration refers to another version, the YD RP2040 Lite, which has only 2MB of flash.) In my case, the setup of 240Mhz@W25Q080/4 works well.

**Reply by maxgerhardt** [reply]
When I search "YD RP2040" I get results for 4MByte, 8Mbyte and 16Mbyte.. but indeed none 2Mbyte. Can you file an issue for that we have proper board defintions for YD RP2040 (4,8,16) and YD RP2040 Lite?

In terms of the problems, the `board_build.f_cpu = 240000000L` line should work for an overclock to 240MHz, it's [the same as Arduino IDE](https://github.com/earlephilhower/arduino-pico/blob/3cc5ac14ff5ab9fd2d0399c8d4a99c3451c5e818/boards.txt#L100-L125) does. LittleFS may however be bottlenecked by flash read time instead of computation time.  I see you tried to counteract that by giving it another Boot2 file, which should work. ESP32-style `flash_mode` and `f_flash` instructions are not the correct configuration path, that's what the Boot2 file is for. Which problems remain?

**Reply by ws2bhhh** [reply]
```
 Can you file an issue for that we have proper board defintions for YD RP2040 (4,8,16) and YD RP2040 Lite?
```
I'm not a native English speaker, so I can't understand exactly what I'm supposed to do next, I wrote down a few things that might be relevant.
Well, They released their schematic diagram (named as SCH.pdf) at:
YD RP2040: http://152.32.187.208:8080/yd-data/YD-RP2040/
YD RP2040 Lite: http://152.32.187.208:8080/yd-data/YD-RP2040/YD-RP2040-LITE/
If you can't open it, let me know and I'll upload it directly to github.
```
 Which problems remain?
```
My on-board flash model number is ZB25VQ32BSIG, 4m byte.
I guess I also need to fix the problem of slow file reading? Even though I've set the flash speed to 120mhz (by using boot2_w25q080_2_padded_checksum.S), reading still takes more than 12ms.

**Reply by maxgerhardt** [reply, marked as answer]
Flash speed is decided by CPU frequency and Boot2 file. The `_2_` in the Boot2 file name means the code is for a SPI clock frequency divider of "2", i.e., CPU frequency divided by 2. "4" respectively for DIV 4. A Pico with 7ms read time compared to 12ms on an ESP32 seems understandable to me, after all, the Pico is only Cortex M0+ with no hardware integer divider and the code may need a few of those. The ESP32C3's RISC-V (RV32IMC) core is better in regards to that. 

Note: With the ZB25VQ32BSIG, you have
>SPI clock frequencies of up to 120MHz are supported allowing equivalent clock rates of 240MHz (120MHz x 2) for Dual I/O and 416MHz (104MHz x 4) for Quad I/O when using the Fast Read Dual/Quad I/O and QPI instructions

So if you set F_CPU to 240MHz and the Boot2 file with SPI clock divder = 2, then the SPI CLK is already at 120MHz, which is technically out-of-specs for the flash chip already.

You might need a shift in paradigm: Remember that the Pico has **TWO** Cortex-M0+ and can calculate things independendly of each other. Meaning, you could write a program that reads the LittleFS file on one core, and via an IPC (inter-process-communication) like a FreeRTOS queue or shared memory,  let the *second core* act on the recieved data and free it. If the second core needs 5ms processing time to act on the data, it only needs to wait for 7ms for the other core to push it the next piece of data and the cycle can begin again.

**Reply by ws2bhhh** [reply]
littlefs significantly slower than fatfs… In fact, even if I wrote some bad code, reading the same file byte by byte unnecessarily more than three times, the time consumed by fatfs can still be equal to the correctly written littlefs that uses the buffer method to read all at once.
Maybe I should try to port fatfs. Implementing the simple read function should not be very difficult.🤔

**Comment by ws2bhhh** [comment]
For those who suffer from the same read speed problem, I tried to port a read-only fatfs to the rp2040 (**writing seems impossible due to the use of XIP**). Reading a 700 byte file at 200mhz @ divder = 2 takes about **1ms**. For me at least, this is already faster than I thought possible. More tests are still needed to see if reading more than 1024 bytes slows it down.

**Reply by maxgerhardt** [reply]
Very nice, but just FYI, the EEPROM library also writes to flash (it's an EEPROM emulation based on flash), so it should not be impossible: https://github.com/earlephilhower/arduino-pico/blob/master/libraries/EEPROM/src/EEPROM.cpp
